core/src/main/java/org/apache/calcite/rel/rules/UnionPullUpConstantsRule.java

```diff
@@ -0,0 +1,147 @@
+          new UnionPullUpConstantsRule(Union.class, RelFactories.LOGICAL_BUILDER);
+
+  protected UnionPullUpConstantsRule(
+      Class<? extends Union> unionClass,
+      RelBuilderFactory relBuilderFactory) {
+    super(operand(unionClass, any()),
+            relBuilderFactory, null);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Union union = call.rel(0);
+
+    final int count = union.getRowType().getFieldCount();
+    if (count == 1) {
+      // No room for optimization since we cannot create an empty
+      // Project operator.
+      return;
+    }
+
+    final RexBuilder rexBuilder = union.getCluster().getRexBuilder();
+    final RelMetadataQuery mq = RelMetadataQuery.instance();
+    final RelOptPredicateList predicates = mq.getPulledUpPredicates(union);
+    if (predicates == null) {
+      return;
+    }
+
+    Map<RexNode, RexNode> conditionsExtracted = ReduceExpressionsRule.predicateConstants(
+            RexNode.class, rexBuilder, predicates);
+    Map<RexNode, RexNode> constants = new HashMap<>();
+    for (int i = 0; i < count; i++) {
+      RexNode expr = rexBuilder.makeInputRef(union, i);
+      if (conditionsExtracted.containsKey(expr)) {
+        constants.put(expr, conditionsExtracted.get(expr));
+      }
+    }
+
+    // None of the expressions are constant. Nothing to do.
+    if (constants.isEmpty()) {
+      return;
+    }
+
+    // Create expressions for Project operators before and after the Union
+    List<RelDataTypeField> fields = union.getRowType().getFieldList();
+    List<RexNode> topChildExprs = new ArrayList<>();
+    List<String> topChildExprsFields = new ArrayList<>();
+    List<RexNode> refs = new ArrayList<>();
+    ImmutableBitSet.Builder refsIndexBuilder = ImmutableBitSet.builder();
+    for (int i = 0; i < count; i++) {
+      RexNode expr = rexBuilder.makeInputRef(union, i);
+      RelDataTypeField field = fields.get(i);
+      if (constants.containsKey(expr)) {
+        topChildExprs.add(constants.get(expr));
+        topChildExprsFields.add(field.getName());
+      } else {
+        topChildExprs.add(expr);
+        topChildExprsFields.add(field.getName());
+        refs.add(expr);
+        refsIndexBuilder.set(i);
+      }
+    }
+    ImmutableBitSet refsIndex = refsIndexBuilder.build();
+
+    // Update top Project positions
+    final Mappings.TargetMapping mapping =
+            RelOptUtil.permutation(refs, union.getInput(0).getRowType()).inverse();
+    topChildExprs = ImmutableList.copyOf(RexUtil.apply(mapping, topChildExprs));
+
+    // Create new Project-Union-Project sequences
+    final RelBuilder relBuilder = call.builder();
+    for (int i = 0; i < union.getInputs().size(); i++) {
+      RelNode input = union.getInput(i);
+      List<Pair<RexNode, String>> newChildExprs = new ArrayList<>();
+      for (int j = 0; j < refsIndex.cardinality(); j++) {
+        int pos = refsIndex.nth(j);
+        newChildExprs.add(
```

Comment:
I think you can write instead {{for (int j : refsIndex)}}, because ImmutableBitSet is iterable.